Reduce default mempool size in -blocksonly mode

willcl-ark opened the pull request:
Fixes #9526

When `-blocksonly` has been set reduce default mempool size to avoid surprising resource usage via sharing un-used mempool cache space with dbcache.

In comparison to https://github.com/bitcoin/bitcoin/pull/9569 which either set `maxmempool` size to 0 when `-blocksonly` was set or else errored on startup, this change will permit `maxmempool` options being set.

This preserves the current (surprising?) behaviour of having a functional mempool in `-blocksonly` mode, to permit whitelisted peer transaction relay, whilst reducing average runtime memory usage for blocksonly nodes which either use the default settings or have otherwise configured a `maxmempool` size.

To use the previous old defaults node operators can configure their node with: `-blocksonly -maxmempool=300`.

willcl-ark committed:
mempool: Don't share mempool with dbcache in blocksonly

When -blockonly is set, reduce mempool size to 5MB unless -maxmempool
is also set.

See #9569

stickies-v reviewed (commented):
It took me a surprisingly long time to figure out the rationale here. I'll write down my understanding both to have it validated/corrected, as well as hopefully to inform other people looking at this.

---

My understanding:

Users running with `-blocksonly` expect to have a near zero-sized mempool, because it'll only contain transactions broadcasted by their own wallet (disabled by default) or transactions received from whitelisted peers (non-default). As such, when provisioning their system, a user would expect to only have to allocate `dbcache` amount of memory, instead of currently actually requiring `dbcache + maxmempool`, which could lead to stability issues.

Without `-blocksonly`, a user needs to ensure that his system is provisioned to support both the full dbcache and mempool. Since it's provisioned _anyway_, we might as well use the unused mempool space for dbcache, which is especially helpful during IBD.

The misunderstanding for me was that the resource sharing between `dbcache` and `maxmempool` is not a resource optimization (in which case "we would benefit _even more_ with `-blocksonly`, so why disable the optimization?!") but rather an accounting trick. Users running in `-blocksonly` should just increase their `dbcache` by whatever size `maxmempool` is for a similar effect (but not entirely, as explained below)

---

If my understanding is correct, I'm Concept ACK (i.e. making resource usage less surprising) but not yet Approach ACK, because: 
- this would be a performance degradation for users running `-blocksonly` that don't realize they should update their `-dbcache`. Users running `-blocksonly` could get higher default `dbcache` defaults, but that seems messy too.
- users can't really replicate current behaviour even if they're aware of the change. The intuitive solution would be to increase `-dbcache` by `-maxmempool`, however, `dbcache` is shared by multiple caches, [of which the coins cache receives 25-50%](https://github.com/bitcoin/bitcoin/blob/7ef730ca84bd71a06f986ae7070e7b2ac8e47582/src/node/caches.cpp#L28). It seems there's no way to allocate an extra 300MB (assuming default) to just the coins cache without increasing the `dbcache` by a total of 600MB - 1.2GB?

willcl-ark commented:
Thanks for your review at this conceptual stage @stickies-v. Your understanding of the problem and solution matches my own, although I had not fully considered the knock-on effect that:

> users can't really replicate current behaviour even if they're aware of the change ... without increasing the dbcache by a total of 600MB - 1.2GB?

I also concur with you that altering default `dbcache` size for users who enable `blocksonly` seems kind of messy, although it could be made to work.

Therefore I see x options available from here:

1) Keep current behaviour, document it better, and close https://github.com/bitcoin/bitcoin/issues/9526 as wontfix
2) Implement (something like) this change which introduces the performance reduction in IBD for `blocksonly` nodes and document it
3) Add special-casing for `dbcache` size when nodes are run in `blocksonly` mode.

In writing the options out like that, my first preference might actually be 1), followed by 3) and last 2).

stickies-v commented:
> Therefore I see x options available from here:

I think I'm not in favour of 1 & 3. Imo the direction should always be to reduce exceptional behaviour, if the short-term cost of that change is not exorbitant. I think in this case, the cost (potentially slower IBD which can be fixed with user config) is not exorbitant and increased stability (because of reduction of unexpected behaviour) is a worthwhile improvement.

I can think of two other alternatives, both building on top of what you've currently implemented:

4. add a `-increase-coins-cache=<n>` parameter to allow users keeping the current behaviour (and maybe even set it to 300 by default if `-blocksonly` in the first release to give users time to transition, and then change it to 0 in the release after. Pretty user-friendly/straightforward option, but it introduces another kinda one-off parameter.
 
5. Allowing users to manually define sub-cache allocations (overriding the defaults), e.g. through `-dbcache.coins=<n|f>` where `n` is an absolute amount and `f` a 0-1 float indicating a percentage of total `dbcache`. This seems like the most transparent and generic solution, but probably at the cost of reduced user-friendliness. It would also be a significantly bigger code change introducing the new parameters as well as a bunch of validation, and allowing the auto-allocation to deal with partial specification (e.g. only `coins` and `tx_index` are specified but not the other ones)

willcl-ark commented:
Yes 4) is a possible solution.

IMO by the time we hit 5), we're probably almost talking about people who have such an understanding of the system (developers) that they can most likely make the change to the source themselves before compiling if they truely need such fine-grained settings changes.

Will leave it as is to see if we get any other suggestions, and mull it over myself...

willcl-ark commented on src/validation.cpp:
```diff
@@ -2336,8 +2336,16 @@ CoinsCacheSizeState Chainstate::GetCoinsCacheSizeState(
+    int64_t nTotalSpace;
+    // Don't share unused mempool space with dbcache in blocksonly mode to avoid surprising resource usage
+    // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
+    if (gArgs.GetBoolArg("-blocksonly", false)) {
```
This is in net.h, which I don't really want to add as a new dependency for validation.cpp 🤔

stickies-v replied:
Fair enough, but then I think `DEFAULT_BLOCKSONLY` should be declared in a different header (not too sure where that would be, hopefully someone more familiar can chime in). Redeclaring default values in multiple locations scares me.

dergoegge replied:
Ideally we should probably remove Chainstate's dependency on `ArgsManager` and introduce some chainstate options. That should probably happen in a separate PR though.

willcl-ark replied:
Right. Have taken a new direction though for now as per https://github.com/bitcoin/bitcoin/pull/26471#discussion_r1072155280

luke-jr commented:
> The intuitive solution would be to increase -dbcache by -maxmempool, however, dbcache is shared by multiple caches, of which the coins cache receives 25-50%. It seems there's no way to allocate an extra 300MB (assuming default) to just the coins cache without increasing the dbcache by a total of 600MB - 1.2GB?

FWIW, I have a related thought process for Knots, where it makes sense to leave the coins cache unlimited (and let memory pressure trigger flushing), yet it isn't sane to just increase `-dbcache` to the system memory size, because that would also increase other cache sizes which aren't handled when under memory pressure.

>Therefore I see x options available from here:

6) support specifying "+mempool" on the end of the `-dbcache` option to get the current behaviour even in blocksonly mode.

willcl-ark commented:
> 7. Perhaps setting a very small mempool `m_max_size_bytes` when in blocksonly mode unless overridden by `-maxmempool`, would be one solution? That seems to preserve the trick for those using it intentionally, while reducing the amount of "surprising memory usage" for those that assume 0 memory for `-blocksonly`. Those who benefit from the extra 300MB by accident can perhaps get their answer from the release notes if they see a performance reduction. If they want to continue doing what they were doing, they just need to set `-maxmempool`.

I think this is a neat solution which covers all requirements so implemented it [in this second branch](https://github.com/willcl-ark/bitcoin/commit/2920558e39f2e874c2a53f6b6a284f90a1251dfb) (it's also much cleaner implementation-wise).

The only question I had was how small is "smaller"? I opted for 5MB in blocksonly mode as this should be enough for quite a few transactions to be relayed for whitelisted peers, I think...

If this is preferred I will move it over to this branch...?

willcl-ark committed:
doc: Update blocksonly behaviour in reduce-memory

Changes to the default mempool allocation size now documented.

Provides users with guidance on the mempool implications of -blocksonly
mode, along with instructions on how to re-enable old behaviour.

glozow commented:
> The only question I had was how small is "smaller"? I opted for 5MB in blocksonly mode as this should be enough for quite a few transactions to be relayed for whitelisted peers, I think...

That definitely seems sufficient, and they can still make it higher if they need to. 5MB is the minimum `-maxmempool` config allowed as well, as the smaller the mempool size, the easier it is for your mempool min feerate to get driven up really high. Thinking out loud I guess this isn't as much of a concern for `-blocksonly` since you won't consider any transactions from strangers, so maybe smaller would be okay... but I can't say a different number with confidence.

> If this is preferred I will move it over to this branch...?

I think it's a relatively simple change to lessen this problem with minimal side effects? :shrug: though I don't know if we could say we've completely fixed it. fwiw I think it's sensible regardless since we're probably unlikely to need 300MB in blocksonly mode.

willcl-ark committed:
doc: release note on mempool size in -blocksonly

Adds a release note detailing the new mempool sizing behaviour when
running in blocksonly mode, and instruction on how to override the new
defaults.

willcl-ark commented:
OK I have moved the commits across, and updated the PR title and description. I also added a new release note commit which I thought was worth it, but happy to be persuaded otherwise.

aureleoules commented on src/init.cpp:
```diff
@@ -1467,7 +1467,16 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
+    if (args.GetBoolArg("-blocksonly", 0) && !args.IsArgSet("-maxmempool")) {
+        // Re-use descendant_limit_bytes as the minimum mempool size
+        mempool_opts.max_size_bytes = descendant_limit_bytes;
```
nit: `descendant_limit_bytes` is around 3.9MB and not 5MB as claimed. Should the documentation be changed to reflect this or it doesn't really matter?

aureleoules commented on doc/reduce-memory.md:
```diff
@@ -16,10 +16,14 @@ The size of some in-memory caches can be reduced. As caches trade off memory usa
 
   - Do not use this when using the client to broadcast transactions as any transaction sent will stick out like a sore thumb, affecting privacy. When used with the wallet it should be combined with `-walletbroadcast=0` and `-spendzeroconfchange=0`. Another mechanism for broadcasting outgoing transactions (if any) should be used.
 
+  - In `-blocksonly` mode the node will by default use a 5MB mempool size, avoiding unexpected memory usage from dbcache sharing the 300MB default mempool cache size.
+
+  - If you would like to continue using the mempool cache allocation for dbcache in `-blocksonly` mode then you can specify the mempool size using the `maxmempoolsize` option.
```
nit: Maybe also add a note saying that setting `-limitdescendantsize` can also alter the cache size with `-blocksonly`?

`./src/bitcoind -limitdescendantsize=0`
> Using 440.0 MiB for in-memory UTXO set (plus up to 286.1 MiB of unused mempool space)

`./src/bitcoind -limitdescendantsize=300`
> Using 440.0 MiB for in-memory UTXO set (plus up to 286.1 MiB of unused mempool space)

---
`./src/bitcoind -limitdescendantsize=0 -blocksonly`
> Using 440.0 MiB for in-memory UTXO set (plus up to 0.0 MiB of unused mempool space)

`./src/bitcoind -limitdescendantsize=300 -blocksonly`
> Using 440.0 MiB for in-memory UTXO set (plus up to 11.4 MiB of unused mempool space)

willcl-ark replied:
Hmmm you are right. I wanted to avoid another magic constant if possible, but this does indeed have unintended side effects. I am going to revert it to 5MB for now

willcl-ark commented on src/init.cpp:
```diff
@@ -1467,7 +1467,16 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    // If we are running in blocksonly mode don't allocate the default 300MB of space for mempool
+    // unless the user specifies -maxmempool, as otherwise it will leak into dbcache allowance.
+    // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
+    if (args.GetBoolArg("-blocksonly", 0) && !args.IsArgSet("-maxmempool")) {
```
Ah yes indeed. Thanks!

aureleoules reviewed (approved):
ACK 931924f5cc1dd9c82306a0ce59b62cf5c9276565
Tested with `-blocksonly` and `-blocksonly -maxmempool=n` and verified the logs, seems to work as intended.

(Please avoid rebasing when there is no need.)

stickies-v commented on doc/reduce-memory.md:
```diff
@@ -16,10 +16,14 @@ The size of some in-memory caches can be reduced. As caches trade off memory usa
 
+  - In `-blocksonly` mode the node will by default use a 5MB mempool size, avoiding unexpected memory usage from dbcache sharing the 300MB default mempool cache size.
+
+  - If you would like to continue using the mempool cache allocation for dbcache in `-blocksonly` mode then you can specify the mempool size using the `maxmempoolsize` option.
```
```suggestion
  - If you would like to continue using the mempool cache allocation for dbcache in `-blocksonly` mode then you can specify the mempool size using the `-maxmempool` option.
```

willcl-ark replied:
Whoops, fixed.

stickies-v commented on doc/release-notes-26471.md:
```diff
@@ -0,0 +1,14 @@
+Updated settings
+----------------
+
+- Previously running with the `-blocksonly` option effectively disabled the
```
```suggestion
- Previously running with the `-blocksonly` option all but disabled the
```

willcl-ark replied:
Personally I feel "effectively disabled" is cleared here, so I've left it as-is for now (but happy to be persuaded)

stickies-v replied:
The definition "actually but not officially or explicitly" captures "effectively" quite well for me, and in that sense I think is wrong here. But won't comment further, not that important.

willcl-ark replied:
Removed in new version

stickies-v commented on doc/release-notes-26471.md:
```diff
@@ -0,0 +1,14 @@
+
+- Previously running with the `-blocksonly` option effectively disabled the
+  mempool but continued to allocate 300MB of cache for it, which would be
+  shared with the dbcache. This might lead to unexpected resource usage of
+  nodes running with this option during for example IBD.
```
```suggestion
  shared with the dbcache. This might lead to nodes running with this
  option using more resources than anticipated, especially during IBD.
```

willcl-ark replied:
Done

stickies-v commented on doc/release-notes-26471.md:
```diff
@@ -0,0 +1,14 @@
+
+  In this release running with `-blocksonly` will only allocate 5MB to the
+  mempool by default, permitting the use-case of relaying transactions from
+  whitelisted peers, but avoiding unexpected resource usage.
```
```suggestion
  whitelisted peers, but avoiding unexpected resource usage. It reduces the
  available `dbcache` size by 295MB, impacting performance especially during
  IBD.
```

willcl-ark replied:
Done

stickies-v commented on src/init.cpp:
```diff
@@ -1467,7 +1467,15 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     }
-    LogPrintf("* Using %.1f MiB for in-memory UTXO set (plus up to %.1f MiB of unused mempool space)\n", cache_sizes.coins * (1.0 / 1024 / 1024), mempool_opts.max_size_bytes * (1.0 / 1024 / 1024));
+    // If we are running in blocksonly mode don't allocate the default 300MB of space for mempool
+    // unless the user specifies -maxmempool, as otherwise it will leak into dbcache allowance.
```
In quite a few places we refer to mempool cache being added to dbcache, but it's added specifically to the coins/UTXO cache so I think our language should reflect that consistently? (here and in other places)

```suggestion
    // unless the user specifies -maxmempool, as otherwise it will leak into UTXO cache allowance.
```

willcl-ark replied:
Agreed and changed.

stickies-v commented on src/init.cpp:
```diff
@@ -1467,7 +1467,15 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    // If we are running in blocksonly mode don't allocate the default 300MB of space for mempool
+    // unless the user specifies -maxmempool, as otherwise it will leak into dbcache allowance.
+    // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
+    if (args.GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY) && !args.IsArgSet("-maxmempool")) {
+        mempool_opts.max_size_bytes = 5 * 1'000'000;
```
Have you considered adding this logic to `mempool_args.cpp::ApplyArgsManOptions()` instead, so we don't set our `mempool_opts.max_size_bytes` in two different locations? I think this has the benefit of 1) being easier to understand why options are set to the values they are and 2) avoid unexpected issues/behaviour where `mempool_opts.max_size_bytes` is temporarily set to 300MB before being changed to 5MB.

The downside, [as previously discussed](https://github.com/bitcoin/bitcoin/pull/26471#discussion_r1020227691), is including `net.h`, which probably needs fixing anyway (perhaps in separate PR?).

I think I'd much prefer this approach.

<details>
<summary>git diff on 931924f5cc1dd9c82306a0ce59b62cf5c9276565</summary>

```diff
diff --git a/src/init.cpp b/src/init.cpp
index c4b15392c..641fef81c 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1467,12 +1467,6 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     if (mempool_opts.max_size_bytes < 0 || mempool_opts.max_size_bytes < descendant_limit_bytes) {
         return InitError(strprintf(_("-maxmempool must be at least %d MB"), std::ceil(descendant_limit_bytes / 1'000'000.0)));
     }
-    // If we are running in blocksonly mode don't allocate the default 300MB of space for mempool
-    // unless the user specifies -maxmempool, as otherwise it will leak into dbcache allowance.
-    // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
-    if (args.GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY) && !args.IsArgSet("-maxmempool")) {
-        mempool_opts.max_size_bytes = 5 * 1'000'000;
-    }
     LogPrintf("* Using %.1f MiB for in-memory UTXO set (plus up to %.1f MiB of unused mempool space)\n",
               cache_sizes.coins * (1.0 / 1024 / 1024),
               mempool_opts.max_size_bytes * (1.0 / 1024 / 1024));
diff --git a/src/node/mempool_args.cpp b/src/node/mempool_args.cpp
index 8c929e5e0..b530ac981 100644
--- a/src/node/mempool_args.cpp
+++ b/src/node/mempool_args.cpp
@@ -10,6 +10,7 @@
 #include <chainparams.h>
 #include <consensus/amount.h>
 #include <logging.h>
+#include <net.h>
 #include <policy/feerate.h>
 #include <policy/policy.h>
 #include <script/standard.h>
@@ -42,7 +43,14 @@ std::optional<bilingual_str> ApplyArgsManOptions(const ArgsManager& argsman, con
 {
     mempool_opts.check_ratio = argsman.GetIntArg("-checkmempool", mempool_opts.check_ratio);
 
-    if (auto mb = argsman.GetIntArg("-maxmempool")) mempool_opts.max_size_bytes = *mb * 1'000'000;
+    if (auto mb = argsman.GetIntArg("-maxmempool")) {
+        mempool_opts.max_size_bytes = *mb * 1'000'000;
+    } else if (argsman.GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY)) {
+        // If we are running in blocksonly mode don't allocate the default 300MB of space for mempool
+        // unless the user specifies -maxmempool, as otherwise it will leak into dbcache allowance.
+        // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
+        mempool_opts.max_size_bytes = 5 * 1'000'000;
+    }
 
     if (auto hours = argsman.GetIntArg("-mempoolexpiry")) mempool_opts.expiry = std::chrono::hours{*hours};
 ```
</details>

stickies-v replied:
The docstring can also be simplified now btw, since the described behaviour is in the same if/else

willcl-ark replied:
Moved logic into `mempool_args.cpp::ApplyArgsManOptions()` as this seems both cleaner, and a more logical place for it.

Adding net.h to mempool_args.cpp increased translation unit line count by 29 lines

stickies-v reviewed (commented):
Approach ACK 931924f5cc1dd9c82306a0ce59b62cf5c9276565, [except](https://github.com/bitcoin/bitcoin/pull/26471#discussion_r1072155280) for setting mempool_opts.max_size_bytes in two different locations and times.

I think this approach tackles a nice balance between addressing the most important issue (node crashing because of unexpected resource usage), reducing additional code overhead (compared to the other approaches considered) and allowing users to maintain existing performance profile without too much admin overhead. The biggest (but manageable) downside is that everyone that upgrades and doesn't manually override `-maxmempool` will silently experience a performance degradation, especially during IBD. I expect this performance degradation to be manageable, especially since most users that have the resources available, will probably have manually increased their `dbcache` anyway for IBD, lowering the relative impact of this change.

willcl-ark commented on doc/reduce-memory.md:
```diff
@@ -16,10 +16,14 @@ The size of some in-memory caches can be reduced. As caches trade off memory usa
   - A lower maximum mempool size means that transactions will be evicted sooner. This will affect any uses of `bitcoind` that process unconfirmed transactions.
 
-- To completely disable mempool functionality there is the option `-blocksonly`. This will make the client opt out of receiving (and thus relaying) transactions completely, except as part of blocks.
+- To disable mempool functionality there is the option `-blocksonly`. This will make the client opt out of receiving (and thus relaying) transactions, except from whitelisted peers and as part of blocks.
```
Done

willcl-ark commented on doc/reduce-memory.md:
```diff
@@ -16,10 +16,14 @@ The size of some in-memory caches can be reduced. As caches trade off memory usa
 
   - Do not use this when using the client to broadcast transactions as any transaction sent will stick out like a sore thumb, affecting privacy. When used with the wallet it should be combined with `-walletbroadcast=0` and `-spendzeroconfchange=0`. Another mechanism for broadcasting outgoing transactions (if any) should be used.
 
+  - In `-blocksonly` mode the node will by default use a 5MB mempool size, avoiding unexpected memory usage from dbcache sharing the 300MB default mempool cache size.
```
Done (for now, pending including net.h)

willcl-ark commented on src/init.cpp:
```diff
@@ -1467,7 +1467,15 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    // unless the user specifies -maxmempool, as otherwise it will leak into dbcache allowance.
+    // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
+    if (args.GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY) && !args.IsArgSet("-maxmempool")) {
+        mempool_opts.max_size_bytes = 5 * 1'000'000;
```
Done (as `DEFAULT_BLOCKSONLY_MAX_MEMPOOL_SIZE_MB`)

aureleoules reviewed (approved):
reACK ed5ce6d0c911ac66bd5870a1711edef03a18ebeb - Wording updated and logic cleaned up since my last review.

stickies-v commented on src/kernel/mempool_options.h:
```diff
@@ -18,6 +18,8 @@ class CBlockPolicyEstimator;
 
 /** Default for -maxmempool, maximum megabytes of mempool memory usage */
 static constexpr unsigned int DEFAULT_MAX_MEMPOOL_SIZE_MB{300};
+//** Defaul for -maxmempool when blocksonly is set */
```
```suggestion
/** Default for -maxmempool when blocksonly is set */
```

willcl-ark replied:
Fixed

stickies-v commented on doc/release-notes-26471.md:
```diff
@@ -0,0 +1,16 @@
+  In this release running with `-blocksonly` will only allocate 5MB to the
+  mempool by default, permitting the use-case of relaying transactions from
+  whitelisted peers, but avoiding unexpected resource usage. It reduces the
+  available `dbcache` size by 295MB, impacting performance especially during
```
Perhaps need to add here that this is only true if `-maxmempool` is not set? We say it in other places, but better to not assume everyone reads everything, so seems like a safe option.

stickies-v reviewed (approved):
ACK ed5ce6d0c

ajtowns reviewed (commented):
Approach ACK

stickies-v commented on doc/reduce-memory.md:
```diff
@@ -16,11 +16,15 @@ The size of some in-memory caches can be reduced. As caches trade off memory usa
   - Do not use this when using the client to broadcast transactions as any transaction sent will stick out like a sore thumb, affecting privacy. When used with the wallet it should be combined with `-walletbroadcast=0` and `-spendzeroconfchange=0`. Another mechanism for broadcasting outgoing transactions (if any) should be used.
 
-- Since `0.14.0`, unused memory allocated to the mempool (default: 300MB) is shared with the UTXO cache, so when trying to reduce memory usage you should limit the mempool, with the `-maxmempool` command line argument.
+  - In `-blocksonly` mode the node will by default use a 5MB (`DEFAULT_BLOCKSONLY_MEMPOOL_MAX_SIZE_MB`) mempool size, avoiding unexpected memory usage from the UTXO cache sharing the 300MB default mempool cache size.
```
I agree it looks a bit awkward but I suggested it to make it easier to keep the docs up to date when the constant changes in the future: https://github.com/bitcoin/bitcoin/pull/26471#discussion_r1072128199

No strong view either way.

willcl-ark replied:
Constant removed

stickies-v commented on src/kernel/mempool_options.h:
```diff
@@ -18,6 +18,8 @@ class CBlockPolicyEstimator;
 
 /** Default for -maxmempool, maximum megabytes of mempool memory usage */
 static constexpr unsigned int DEFAULT_MAX_MEMPOOL_SIZE_MB{300};
+//** Defaul for -maxmempool when blocksonly is set */
```
beat ya to it https://github.com/bitcoin/bitcoin/pull/26471#discussion_r1081618786

willcl-ark replied:
Done

willcl-ark commented on src/node/mempool_args.cpp:
```diff
@@ -42,7 +43,14 @@ std::optional<bilingual_str> ApplyArgsManOptions(const ArgsManager& argsman, con
+        // If we are running in blocksonly mode don't allocate the default 300MB of space for mempool
+        // unless the user specifies -maxmempool, as otherwise it will leak into coins/UTXO cache allowance.
+        // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
+        mempool_opts.max_size_bytes = DEFAULT_BLOCKSONLY_MAX_MEMPOOL_SIZE_MB * 1'000'000;
```
Thanks! This both makes more sense, and avoids needing to include _net.h_ in _mempool_args.cpp_, which I didn't particularly like

willcl-ark commented on doc/reduce-memory.md:
```diff
@@ -16,11 +16,15 @@ The size of some in-memory caches can be reduced. As caches trade off memory usa
   - Do not use this when using the client to broadcast transactions as any transaction sent will stick out like a sore thumb, affecting privacy. When used with the wallet it should be combined with `-walletbroadcast=0` and `-spendzeroconfchange=0`. Another mechanism for broadcasting outgoing transactions (if any) should be used.
 
-- Since `0.14.0`, unused memory allocated to the mempool (default: 300MB) is shared with the UTXO cache, so when trying to reduce memory usage you should limit the mempool, with the `-maxmempool` command line argument.
+  - In `-blocksonly` mode the node will by default use a 5MB (`DEFAULT_BLOCKSONLY_MEMPOOL_MAX_SIZE_MB`) mempool size, avoiding unexpected memory usage from the UTXO cache sharing the 300MB default mempool cache size.
+
+  - If you would like to continue using the mempool cache allocation for the UTXO cache in `-blocksonly` mode (to help speed up IBD) then you can specify the mempool size using the `-maxmempool` option.
```
Agreed and done

willcl-ark commented on doc/reduce-memory.md:
```diff
@@ -16,11 +16,15 @@ The size of some in-memory caches can be reduced. As caches trade off memory usa
-- To completely disable mempool functionality there is the option `-blocksonly`. This will make the client opt out of receiving (and thus relaying) transactions completely, except as part of blocks.
+- Since `0.14.0`, unused memory allocated to the mempool (default: 300MB) is shared with the UTXO cache, so when trying to reduce memory usage you should limit the mempool, with the `-maxmempool` command line argument.
+
+- To disable most of the mempool functionality there is the `-blocksonly` option. This will make the client opt out of receiving (and thus relaying) transactions, except from whitelisted peers and as part of blocks.
```
Done

willcl-ark commented on src/node/mempool_args.cpp:
```diff
@@ -42,7 +43,14 @@ std::optional<bilingual_str> ApplyArgsManOptions(const ArgsManager& argsman, con
+    if (auto mb = argsman.GetIntArg("-maxmempool")) {
+        mempool_opts.max_size_bytes = *mb * 1'000'000;
+    } else if (argsman.GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY)) {
+        // If we are running in blocksonly mode don't allocate the default 300MB of space for mempool
+        // unless the user specifies -maxmempool, as otherwise it will leak into coins/UTXO cache allowance.
+        // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
```
Removed text as part of using AJ's suggestion with `InitParameterInteraction`. I think combined with git log it should be easy enough for people to find out why this interaction was added.

willcl-ark commented on doc/release-notes-26471.md:
```diff
@@ -0,0 +1,16 @@
+Updated settings
+----------------
+
+- Previously running with the `-blocksonly` option effectively disabled the
+  mempool but continued to allocate 300MB of cache for it, which would be
+  shared with the UTXO cache. This might lead to nodes running with this
+  option using more resources than anticipated, especially during IBD.
+
+  In this release running with `-blocksonly` will only allocate 5MB to the
+  mempool by default, permitting the use-case of relaying transactions from
+  whitelisted peers, but avoiding unexpected resource usage. It reduces the
+  available `dbcache` size by 295MB, impacting performance especially during
+  IBD.
```
I agree that we don't  need to mention the whitelist relay behaviour, but its apparently something some people (might) [have been doing](https://github.com/bitcoin/bitcoin/pull/9569#issuecomment-273338305).

Anyway, removed/

willcl-ark commented:
Thanks for the comments, I've updated the branch taking them into consideration.

ajtowns reviewed (approved):
ACK 8e85164e7d12be324ea1af2e288ebcf689c930b7

stickies-v reviewed (approved):
re-ACK https://github.com/bitcoin/bitcoin/commit/8e85164e7d12be324ea1af2e288ebcf689c930b7

fanquake commented on src/kernel/mempool_options.h:
```diff
@@ -18,6 +18,8 @@ class CBlockPolicyEstimator;
 
 /** Default for -maxmempool, maximum megabytes of mempool memory usage */
 static constexpr unsigned int DEFAULT_MAX_MEMPOOL_SIZE_MB{300};
+//** Default for -maxmempool when blocksonly is set */
```
```suggestion
/** Default for -maxmempool when blocksonly is set */
```

MarcoFalke commented on doc/reduce-memory.md:
```diff
@@ -16,11 +16,11 @@ The size of some in-memory caches can be reduced. As caches trade off memory usa
+- Since `0.14.0`, unused memory allocated to the mempool (default: 300MB) is shared with the UTXO cache, so when trying to reduce memory usage you should limit the mempool, with the `-maxmempool` command line argument.
 
-  - Do not use this when using the client to broadcast transactions as any transaction sent will stick out like a sore thumb, affecting privacy. When used with the wallet it should be combined with `-walletbroadcast=0` and `-spendzeroconfchange=0`. Another mechanism for broadcasting outgoing transactions (if any) should be used.
+- To disable most of the mempool functionality there is the `-blocksonly` option. This will reduce the default memory usage to 5MB and make the client opt out of receiving (and thus relaying) transactions, except from whitelisted peers and as part of blocks.
```
nit: Instead of whitelisted, it would be good to mention the exact permission flag.

glozow commented:
@willcl-ark will you be taking the followups https://github.com/bitcoin/bitcoin/pull/26471#discussion_r1083792487 and https://github.com/bitcoin/bitcoin/pull/26471#discussion_r1083469754?

willcl-ark commented:
Yes I can do those @glozow
